File: linux-binary-debugging/examples/fibs-hs/callfib.c
#include <stdio.h>
#include <stdint.h>

extern uint64_t fib(uint64_t);
extern uint64_t fibsum(uint64_t,uint64_t);

int main() {
  uint64_t fib1 = 9;
  uint64_t fib2 = 11;
  uint64_t a = fib(fib1);
  printf("fib(%lu) = %lu\n", fib1, a);
  uint64_t b = fib(fib2);
  printf("fib(%lu) = %lu\n", fib2, b);
  uint64_t s = fibsum(fib1,fib2);
  printf("%lu + %lu = %lu\n", a, b, s);
  return 0;
}

File: linux-binary-debugging/examples/fibs-hs/CallFib.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Main where
import Text.Printf
import Foreign.C.Types

foreign import ccall "fib" fib' :: CULong -> CULong
foreign import ccall "fibsum" fibSum' :: CULong -> CULong -> CULong

fib = fromIntegral . fib' . fromIntegral
fibSum a b = fromIntegral $ fibSum' (fromIntegral a) (fromIntegral b)

index1, index2, fib1, fib2 :: Int
index1 = 9
index2 = 11
fib1 = fib index1
fib2 = fib index2

main :: IO ()
main = do
  let s = fib1 + fib2
      s' :: Integer
      s' = fibSum index1 index2

  putStrLn $ printf "fib(%d) = %d" index1 fib1
  putStrLn $ printf "fib(%d) = %d" index2 fib2
  putStrLn $ printf "We calculated:  %d + %d = %d" fib1 fib2 s
  putStrLn $ printf "Lib calculated: %d + %d = %d" fib1 fib2 s'